lib/providers/update_provider.dart

```diff
@@ -0,0 +1,372 @@
+
+    downloading = true;
+    notifyListeners();
+
+    String fileName;
+    String extension;
+
+    if (Platform.isWindows) {
+      fileName = UpdateVersion.windowsDownloadFileName;
+      extension = '.exe';
+    } else if (Platform.isLinux) {
+      assert(linuxEnvironment != null);
+      fileName = UpdateVersion.linuxDownloadFileName;
+      extension = '.$linuxEnvironment';
+    } else {
+      downloading = false;
+      notifyListeners();
+      throw UnsupportedError(
+        'Unsupported platform: ${Platform.operatingSystem}',
+      );
+    }
+
+    final file = File(path.join(tempDir, '$fileName-$version$extension'));
+    if (await file.exists()) await file.delete();
+
+    final executablePath = Uri.https(
+      'github.com',
+      '/bluecherrydvr/unity/releases/download/bleeding_edge/$fileName$extension',
+    );
+
+    await Dio().downloadUri(
+      executablePath,
+      file.path,
+      onReceiveProgress: (received, total) {
+        downloadProgress = received / total * 100;
+        notifyListeners();
+      },
+    );
+
+    downloading = false;
+    downloadProgress = 0.0;
+    notifyListeners();
+  }
+
+  /// Installs the executable for the latest version.
+  ///
+  /// It can not downgrade
+  Future<void> install({
+    required VoidCallback onFail,
+  }) async {
+    assert(isDesktop, 'This should never be reached on non-desktop platforms');
+
+    // TODO(bdlukaa): Introduce this on ready-to-review PR
+    // assert(latestVersion.version != packageInfo.version, 'Already up to date');
+
+    final executable = executableFor(latestVersion.version);
+    assert(executable != null, 'Executable not found');
+
+    windowManager.hide();
+
+    if (Platform.isWindows) {
+      // https://jrsoftware.org/ishelp/index.php?topic=technotes
+      Process.run(executable!.path, [
+        '/SP-',
+        '/silent',
+        '/noicons',
+      ]);
+    } else if (Platform.isLinux) {
+      switch (linuxEnvironment) {
+        case UpdateVersion.rpm:
+          Process.run('tar', ['-U', executable!.path]);
+          break;
+        case UpdateVersion.deb:
+          Process.run('sudo', ['dpkg', '-i', executable!.path]);
+          break;
+        case UpdateVersion.tarball: // tarball
+          Process.run('tar', ['-i', executable!.path]);
+          break;
+        case UpdateVersion.appImage:
+          throw UnsupportedError('AppImages do not support updating from app');
+        default:
+          throw UnsupportedError(
+            'Can not install an executable on an unknown environment',
+          );
+      }
+    }
+
+    windowManager.close();
+  }
+
+  /// Check for new updates.
+  Future<void> checkForUpdates() async {
+    loading = true;
+    notifyListeners();
+
+    tempDir = (await getTemporaryDirectory()).path;
+
+    final versions = <UpdateVersion>[];
+    // Parse the versions from the server
+    final doc = XmlDocument.parse((await http.get(Uri.parse(appCastUrl))).body);
```

Comment:
This can be done on another thread

Reply:
I don't think this is necessary for now. It may be re-evaluated in the future when there are a lot of versions